docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,167 @@
+(much) less custom code for both the server (i.e., the application) and the
+client. The project is well-funded and widely-used, which makes it a safe bet
+for a dependency.
+
+## Decision
+
+There is a set of related alternatives to consider:
+
+- Question 1: Designate a single IPC standard for out-of-process applications?
+
+  Claim: We should converge on one (and only one) IPC option for out-of-process
+  applications. We should choose an option that, after a suitable period of
+  deprecation, can address most or all the highest-impact uses of Tendermint.
+  Maintaining multiple options increases the surface area for bugs and
+  vulnerabilities, and we should not have multiple options for basic interfaces
+  without a clear and well-documented reason.
+
+- Question 2a: Standardize on gRPC and deprecate/remove the socket protocol?
+
+  Claim: Maintaining and improving a custom RPC protocol is a substantial
+  project and not directly relevant to the requirements of consensus. We would
+  be better served by depending on a well-maintained open-source library like
+  gRPC.
+
+- Question 2b: Improve the socket protocol and deprecate/remove gRPC?
+
+  Claim: If we find meaningful advantages to maintaining our own custom RPC
+  protocol in Tendermint, we should treat it as a first-class project within
+  the core and invest in making it good enough that we do not require other
+  options.
+
+## Detailed Design
+
+1. If we choose to standardize on gRPC, the main work in in Tendermint core
+   will be removing and cleaning up the code for the socket client and server.
+
+   Besides the code cleanup, we will also need to clearly document a
+   deprecation schedule, and invest time in making the migration easier for
+   applications currently using the socket protocol.
+
+3. If we choose to keep only the socket protocol, we will need to follow up
+   with a more detailed design for extending and upgrading the protocol to fix
+   the existing performance and operational issues with the protocol.
+
+3. If we choose to keep both options, we will still need to do all the work of
+   (2), but the gRPC implementation should not require any immediate changes.
+
+
+## Consequences
+
+- **Standardize on gRPC**
+
+   - ✅ Addresses existing performance and operational issues.
+   - ✅ Replaces custom code with a well-maintained widely-used library.
+   - ✅ Aligns well with Cosmos SDK, which already uses gRPC extensively.
+   - ✅ Aligns well priv validator interface, for which the socket protocol is already deprecated for gRPC.
+   - ❓ Applications will be hard to implement in a language without gRPC support.
+   - ⛔ All users of the socket protocol have to migrate to gRPC.
+
+- **Standardize on socket protocol**
+
+   - ✅ Less immediate impact for existing users (but see below).
+   - ✅ Simplifies ABCI API surface by removing gRPC.
+   - ❓ Users of the socket protocol will have a (smaller) migration.
```

Comment:
This a crucial point. If we want to keep both, there is still a migration that users will need to perform to use whatever upgraded client we create. Upgrading the socket client also places the cost of implementing the protocol onto non-Go socket protocol library maintainers (of which we are not). This means that anyone wanting to use Tendermint in an out-of-process style is gated by these updates.

That point highlights an additional consequence not listed here: _anyone not using Go or the few supported languages cannot use Tendermint_. While the number of languages w/ gRPC support is limited, the number of languages with Tendermint socket client implementations is far more limited. We immediately lose language support for almost 10+ languages if we maintain only the socket protocol impl. I think it would be worth adding the list of currently supported languages for each protocol in this doc (if one can be found for the socket-client).

Reply:
That's a very good point. I'll add that for sure.

Reply:
I added another paragraph to the Decision section to call out this consideration more explicitly.

Reply:
I keep looking for langauges that I think might actually be used that aren't supported by gRPC, and I think the answer to that is "scheme", everything else basically is (e.g. Haskell, Common Lisp, Erlang, Elixir, ...) so I think we're deeply into a theoretical requirement (broad language support,) for a use case that doesn't entirely exist, but I think it's good to outline things around this anyway.

Reply:
> I keep looking for langauges that I think might actually be used that aren't supported by gRPC

I thought the request here was for a list of the languages with socket protocol support.

Reply:
I added an [appendix](https://github.com/tendermint/tendermint/blob/mjf/unsocketing/docs/architecture/adr-082-socket-protocol.md#appendix-known-implementations-of-abci-socket-protocol) containing a bunch of ABCI server implementations I found. I can't speak to quality, but I filtered out some more fragmentary examples by visual inspection. These are the ones that show some evidence of maybe implementing most/all of the protocol.

Reply:
> I think that Tendermint should be usable in multiple languages but we should focus on 1. popular languages developers interested in blockchains are likely to use and 2. making sure that whatever implementation we chose is well supported.

Agreed.

> Right now 'You can go implement our socket protocol in your language if you want' is the best we can say right now, which I think is a fairly bad UX. 

Also agreed. Perhaps more importantly, anyone who wants to implement an ABCI server is already on the hook for having good protocol buffer support. That's required even if the goal is to implement the socket protocol for transport.

While I'm sure the subset of "languages that have protocol buffer support but not gRPC" is nonempty, I would be very surprised if languages in that subset has any impact on the usage footprint of Tendermint.